### DISCUSSION TITLE
v5 Serialize a `preload()` relation

### DISCUSSION THREAD

**Discussion by ad-on-is** [adonisjs/core#2365] [Help]
I'm trying to serialize user data based on different scenarios.

**/user/1234/products/**
`Product.query().where({user_id: 1234}).preload('users')`
_Should only show name, avatar in the user relation_

**/auth/me**
`User.find(auth.user.id) `
_Should show all user fields_

Unfortunately, `query()` has  no `serialize()` chain and I cannot set `serializeAs` in the model, because then `User.find()` won't show all the fields.

Is there anything I'm missing?

**Comment by RomainLanz** [comment]
Hey @ad-on-is! 👋 

You can pass a callback to the `preload` method to select only some fields.

```ts
Product.query()
  .where({ user_id: 1234 })
  .preload('users', (query) => query.select('name', 'avatar'))
```

**Reply by ad-on-is** [reply]
This kind of works, but it still exposes fields which I'd like to hide. 
For example, it still exposes the `id` field. Also, if I have a `@computed()` field `first_name`, which needs `full_name` to be present in the query, it exposes `full_name` too.

**Reply by RomainLanz** [reply]
You can also customize which field you would like to serialize by passing a callback to `serialize()`.

➡️ https://preview.adonisjs.com/guides/models/serializing-models#cherry-picking-fields

**Reply by ad-on-is** [reply]
Unfortunately there is no method `serialize()` in my situation

```js
const products = Product.query()
  .where({ user_id: 1234 })
  .preload('users')
  .serialize() // Property 'serialize' does not exist on type 'ModelQueryBuilderContract<typeof Product, Product>'

//or
result = await products
result.serialize() // Property 'serialize' does not exist on type 'Product[]'
```

or

```js
Product.query()
  .where({ user_id: 1234 })
  .preload('users', (query) => {query.serialize()}) // Property 'serialize' does not exist on type 'RelationQueryBuilderContract<typeof User, any>'

**Reply by thetutlage** [reply]
The following query returns an array of `Products` and the `serialize` method exists on the `Product` model instance.

```ts
const products = Product.query().where({user_id: 1234}).preload('users')
```

Just loop over the array and serialize as you want.

```ts
products.map((product) => {
  return product.serialize({
    relations: {
      users: {
        fields: ['name', 'avatar']
      }
    }
  })
})
```

**Reply by ad-on-is** [reply]
Ah ok, thanks for that.
